### PR TITLE
fix: Add tests for the AutoTuner and fix bug in _find_nearest_profile

### PR DESCRIPTION
**Pull request by danisereb**
## 📌 Description

I've found a possible bug in the AutoTuner (in function `_find_nearest_profile`).

Before attempting to fix the bug, we need unit test for the AutoTuner to verify its functionality.

The new tests pass (with xfails for the known bug):
```
cd tests/autotuner/

pytest --tb=short .

======================================================== test session starts =========================================================
platform linux -- Python 3.12.3, pytest-8.4.2, pluggy-1.6.0
rootdir: ...
configfile: pytest.ini
plugins: anyio-4.11.0
collected 39 items                                                                                                                   

test_autotuner_bmm_fp8.py ............                                                                                         [ 30%]
test_autotuner_core.py ...........xxxxxx..........                                                                             [100%]
```

The `xxxxxx` are the **xfail** tests.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Tests**
  * Added tests validating FP8 GEMM autotuning: hardware capability checks, autotuner cache behavior across cold/tune/warm scenarios, optional pre-tune flow, and FP8→FP16 correctness with finiteness assertions.
  * Added extensive AutoTuner core tests covering cache management, profile generation and selection, input-preparation behavior, tuning workflows, dynamic bucketing, edge cases, and expected failures to ensure robust autotuner logic.

### CODE REVIEW

File: tests/autotuner/test_autotuner_core.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+    assert all(shape[0] == expected_bucket for shape in nearest)
+    for nearest_shape, original_shape in zip(nearest, shapes, strict=True):
+        assert nearest_shape[1:] == original_shape[1:]
+
+
+@XFAIL_LINKED_DYNAMIC
+def test_find_nearest_profile_moe_same_bucket_same_profile():
+    """MoE inputs mapping to the same bucket should share an identical profile."""
+    config = TuningConfig(
```
[the remark this answers is not in the document]
Comment (danisereb) [reply tests/autotuner/test_autotuner_core.py]: We plan to fix the bug soon and remove the xfail.

File: tests/autotuner/test_autotuner_core.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    )
+    p1 = AutoTuner._find_nearest_profile(_moe_input_shapes(1000), config)
+    p2 = AutoTuner._find_nearest_profile(_moe_input_shapes(1023), config)
+    assert p1 == p2
+
+
+@XFAIL_LINKED_DYNAMIC
+def test_find_nearest_profile_maps_all_linked_dims():
+    """One logical dynamic axis should update every linked tensor/dimension."""
+    tuning_config = TuningConfig(
```
[the remark this answers is not in the document]
Comment (danisereb) [reply tests/autotuner/test_autotuner_core.py]: The test should be fixed along with the bugfix in the AutoTuner class.

File: tests/autotuner/test_autotuner_core.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+@pytest.mark.parametrize(
+    "num_tokens,expected_bucket",
+    [
+        (1024, 1024),
+        (4096, 4096),
+        (8192, 8192),
+        pytest.param(1000, 512, marks=XFAIL_LINKED_DYNAMIC),
+        pytest.param(4000, 2048, marks=XFAIL_LINKED_DYNAMIC),
+        pytest.param(8000, 4096, marks=XFAIL_LINKED_DYNAMIC),
+        pytest.param(12000, 8192, marks=XFAIL_LINKED_DYNAMIC),
```
Comment (danisereb) [inline tests/autotuner/test_autotuner_core.py]: @aleozlx these test cases triggered `using fallback tactic` when I used the trtllm MoE NVFP4 in vLLM.


The `using fallback tactic` happens only when the num tokens is not a power of 2 (I used `FLASHINFER_LOGGING_LEVEL=DEBUG` to see these prints).

**Comment by danisereb** [comment]
@wenscarl can you review this PR ?

File: tests/autotuner/test_autotuner_bmm_fp8.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    _cudnn_gemm_fp8_runner,
+    _get_cache_buf,
+    DEFAULT_WORKSPACE_SIZE,
+)
+from flashinfer.utils import get_compute_capability
+from tests.utils_fp8 import to_float8
+
+
+@pytest.mark.parametrize(
+    "pre_tune,tune_mode,expected_cache_hit",
```
Comment (Naveassaf) [inline tests/autotuner/test_autotuner_bmm_fp8.py:17]: Nit  - would reccommend to have the same order of args and the signature and parameterized.

File: tests/autotuner/test_autotuner_core.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        )
+    )
+
+    shapes: tuple[torch.Size, ...] = _moe_input_shapes(num_tokens)
+
+    nearest = AutoTuner._find_nearest_profile(shapes, config)
+
+    assert all(shape[0] == expected_bucket for shape in nearest)
+    for nearest_shape, original_shape in zip(nearest, shapes, strict=True):
+        assert nearest_shape[1:] == original_shape[1:]
```
[the remark this answers is not in the document]
Comment (danisereb) [reply tests/autotuner/test_autotuner_core.py:155]: Separate test was added

File: tests/autotuner/test_autotuner_bmm_fp8.py (hunk 4 above)
Comment (danisereb) [reply tests/autotuner/test_autotuner_bmm_fp8.py:17]: Fixed

File: tests/autotuner/test_autotuner_core.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                input_idx=(0, 1, 2, 3, 4, 5),
+                dim_idx=(0, 0, 0, 0, 0, 0),
+                gen_tuning_buckets=(512, 1024, 2048, 4096, 8192),
+                map_to_tuning_buckets=lambda x: min(last_positive_power_of_2(x), 8192),
+            ),
+        )
+    )
+    p1 = AutoTuner._find_nearest_profile(_moe_input_shapes(1000), config)
+    p2 = AutoTuner._find_nearest_profile(_moe_input_shapes(1023), config)
+    assert p1 == p2
```
Comment (danisereb) [inline tests/autotuner/test_autotuner_core.py:173]: The values of p1 and p2:
```
test_autotuner_core.py::test_find_nearest_profile_moe_same_bucket_same_profile
p1=((512, 4096), (1000, 256), (1000, 8), (1000, 8), (1000, 4096), (1000, 32))
p2=((512, 4096), (1023, 256), (1023, 8), (1023, 8), (1023, 4096), (1023, 32))
      ^            ^            ^          ^          ^             ^
```

Only the first tuple had its first dim rounded to 512 (last_positive_power_of_2).

The correct result should be:
```
p1=((512, 4096), (512, 256), (512, 8), (512, 8), (512, 4096), (512, 32))
p2=((512, 4096), (512, 256), (512, 8), (512, 8), (512, 4096), (512, 32))
```

The dim0 is the `num_tokens`, the relevant code can be found in flashinfer function `get_trtllm_moe_sm100_module`:

        # their first dimension is num_tokens which will be tuned
        tuning_config_with_hidden_states_scales = TuningConfig(
            dynamic_tensor_specs=(
                DynamicTensorSpec(
                    (0, 1, 2, 3, 4, 5),
                    (0, 0, 0, 0, 0, 0),
                    get_last_power_of_2_num_tokens_buckets(8192, 1),
                    lambda x: min(last_positive_power_of_2(x), 8192),
                    dynamic_tensor_initializers,
                ),
            )
        )

**Comment by danisereb** [comment]
I added the actual fix to `_find_nearest_profile` in the PR (discussed with @wenscarl in slack).

xfail removed from the tests, now all the tests pass:
```
pytest .
================================================= test session starts ==================================================
platform linux -- Python 3.12.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/daserebrenik/workspace/flashinfer
configfile: pytest.ini
plugins: anyio-4.11.0
collected 39 items                                                                                                     

test_autotuner_bmm_fp8.py ............                                                                           [ 30%]
test_autotuner_core.py ...........................                                                               [100%]
============================================ 39 passed, 1 warning in 2.35s =============================================
```

And now we can see the following behavior:
```
test_autotuner_core.py::test_find_nearest_profile_moe_same_bucket_same_profile
p1=((512, 4096), (512, 256), (512, 8), (512, 8), (512, 4096), (512, 32))
p2=((512, 4096), (512, 256), (512, 8), (512, 8), (512, 4096), (512, 32))
```

**Comment by amirkl94** [comment]
@aleozlx

**Comment by wenscarl** [comment]
Maybe update the title of this PR to include words like "fix".

**Comment by danisereb** [comment]
@aleozlx I ran pre-commit:
```
pre-commit run --all-files

check for added large files..............................................Passed
check for case conflicts.................................................Passed
check for merge conflicts................................................Passed
check for broken symlinks............................(no files to check)Skipped
fix end of files.........................................................Passed
mixed line ending........................................................Passed
fix requirements.txt.....................................................Passed
trim trailing whitespace.................................................Passed
Tabs remover.............................................................Passed
CRLF end-lines remover...................................................Passed
clang-format.............................................................Passed
mypy.....................................................................Passed
ruff check...............................................................Passed
ruff format..............................................................Passed
```

I don't see an issue with the new `__init__.py` file it has one empty line, like `tests/gemm/__init__.py`.

No diff between the files:
```
diff autotuner/__init__.py gemm/__init__.py | echo $?
0
```

**Comment by aleozlx** [comment]
tagged as ready to merge
